configfile: "/scratch/group/lilab/Phil/bulkRNASeq_pipeline/config/config.fix.yaml"

## singularity container
 container: "/scratch/group/lilab/Phil/RNASeq_container.sif"

rule all:
    input:
        "02.QC_fastp/QC_summary_table.tsv",
        "02.QC_fastp/duprate.boxplot.png",
#        "03.Alignment_hisat2/mapping.summary",
#        "03.Alignment_hisat2/mappingrate.boxplot.png",
#        "04.Quant_featureCounts/counts_raw.tsv",
#        "04.Quant_featureCounts/counts_filter.tsv",
#        "04.Quant_featureCounts/fc.summary",
#        "04.Quant_featureCounts/assignment_stacked_barplot.png",
#        "05.Normalization_DESeq2/counts_normalized.tsv",
#        expand("05.Normalization_DESeq2/corr.heatmap.{corr_name}.png", corr_name=config["corr"].keys()),
#        expand("05.Normalization_DESeq2/pca.plot.{color_scheme}.png", color_scheme=config["pca_color"].keys()),
#        expand("06.Diff_Expression/{comp}.deseq2_results.tsv", comp=config["diff_comparisons"]),
#        expand("07.Venn/{group}.venn.png", group=config["venn"].keys()),
#        expand("08.Enrichment/{comparison}.go_enrichment.tsv", comparison=config["GO"])

# Include modules
# softlink rawdata 
include: "rules/preprocess.smk"

# rawdata quality control
include: "rules/qc.smk"

## mapping to reference genome
#include: "rules/align.smk"
#
## gene expression quantification
#include: "rules/quant.smk"
#
## row readcount normalization, sample correlation, PCA
#include: "rules/normalize.smk"
#
## differential gene expression
#include: "rules/dge.smk"
#
## plot venn diagram
#include: "rules/venn.smk"
#
## go enrichment analysis
#include: "rules/go.smk"